Nifi 1516 - AWS DynamoDB Get/Put/Delete Processors

mans2singh opened the pull request:
The current implementation supports string and numbers for hash and range keys.  For put operations, a json document can be inserted to a specified field and for get the the json attribute can be retrieved from the item.

jvwing commented:
I believe the unit tests are failing now, for some combination of the following reasons (see the [TravisCI logs](https://s3.amazonaws.com/archive.travis-ci.org/jobs/111920709/log.txt)) 

* ITAbstractDynamoDBTest is no longer marked Ignore 
* The file referenced in CREDENTIALS_FILE does not exist
* Json Document is now required

mans2singh commented:
Hi James:
The unit tests are failing because the credentails file is not found in the home folder.  
DeleteDynamoDBTest.testStringHashStringRangeDeleteNoHashValueFailure:151 Processor has 1 validation failures:
'Credentials File' validated against '/home/travis/aws-credentials.properties' is invalid because File /home/travis/aws-credentials.properties does not exist

My understanding was that the tests marked as IT prefix do not have to be ignored so I had removed that annotation from them.  If that is required, please let me know and I will update them.
Thanks 

    On Friday, February 26, 2016 9:18 AM, James Wing <notifications@github.com> wrote:
 

 I believe the unit tests are failing now, for some combination of the following reasons (see the TravisCI logs)    
   - ITAbstractDynamoDBTest is no longer marked Ignore 
   - The file referenced in CREDENTIALS_FILE does not exist
   - Json Document is now required
—
Reply to this email directly or view it on GitHub.

jvwing commented:
I stand corrected.  The non-integration unit tests are importing CREDENTIALS_FILE from ITAbstractDynamoDBTest, although the integration test itself may not be called.

But I think the unit tests run by TravisCI should pass.  Either the tests requiring a file in the user's home folder are truly integration tests and need to be marked as such, or the file needs to get checked in to git.  I believe you have checked in a similar test credentials file while working on the AWSCredentialsProviderControllerService, and that seemed to work just fine.

It's also possible that the unit tests do not need a credentials file at all, since they won't make API calls to AWS.  Can't we just remove it from the tests?

olegz commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/dynamodb/AbstractDynamoDBProcessor.java:
```diff
@@ -0,0 +1,302 @@
+        }
+    }
+
+    protected DynamoDB getDynamoDB() {
```
Should this by synchronized in some way due to the race condition that exists whenever processor runs as more then one task? Basically there is a chance for DynamoDB instance to be created multiple times.

mans4singh replied:
Made the method synchronized.

mans4singh replied:
Also @jvwing - I've changed the authentication in unit tests to use access key/secret key (dummy values) so I think unit tests should pass now.

mans2singh replied:
@jvwing - Removed creds file from unit tests
@olegz - Synchronized getDynamoDB method

olegz commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/dynamodb/AbstractDynamoDBProcessor.java:
```diff
@@ -0,0 +1,302 @@
+            AmazonServiceException exception) {
+        List<FlowFile> failedFlowFiles = new ArrayList<>();
+        for (FlowFile flowFile : flowFiles) {
+            flowFile = session.putAttribute(flowFile, DYNAMODB_ERROR_EXCEPTION_MESSAGE, exception.getMessage() );
```
It would probably be more efficient to create a Map of attributes and then do _session.putAllAttributes(Map)_ since for each putAttribute a new FlowFile is created.

mans4singh replied:
Updated as you had recommended.

mans2singh replied:
Updated per your recommendation.

olegz commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/dynamodb/DeleteDynamoDB.java:
```diff
@@ -0,0 +1,152 @@
+
+            // All non unprocessed items are successful
+            for (FlowFile flowFile : keysToFlowFileMap.values()) {
+                getLogger().debug("Successfully deleted item to dynamodb : " + table);
```
perhaps deleted "from" not "to"?

mans4singh replied:
Corrected.

mans2singh replied:
Corrected.

olegz commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/dynamodb/DeleteDynamoDB.java:
```diff
@@ -0,0 +1,152 @@
+        } catch(Exception exception) {
+            getLogger().error("Could not process flowFiles due to exception : " + exception.getMessage());
+            List<FlowFile> failedFlowFiles = processException(session, flowFiles, exception);
+            session.transfer(failedFlowFiles, REL_FAILURE);
```
Since handling logic is identical for all 3 exceptions with Exception being the top in the hierarchy can we just remove 
```
catch(AmazonServiceException exception) 
catch(AmazonClientException exception)
```

mans4singh replied:
Amazon service/client exceptions have different attributes (service extends from client). I was using method overloading to process the exceptions accordingly but have not changed the method  names for easy comprehension.

mans2singh replied:
Amazon service/client exception have different attributes (service extends client).  I was using method overloading to populate flow file attribs but have not made the helper methods explicit for comprehension.

olegz commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/dynamodb/GetDynamoDB.java:
```diff
@@ -0,0 +1,180 @@
+        } catch(Exception exception) {
+            getLogger().error("Could not process flowFiles due to exception : " + exception.getMessage());
+            List<FlowFile> failedFlowFiles = processException(session, flowFiles, exception);
+            session.transfer(failedFlowFiles, REL_FAILURE);
```
similar comment as above about exception handling

mans4singh replied:
Same as above - was using method overloading to process specific exception since they have different attributes but have now changed names of process exception method.

mans2singh replied:
Made helper methods explicit.

olegz commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/dynamodb/ItemKeys.java:
```diff
@@ -0,0 +1,52 @@
+/**
+ * Utility class to keep a map of keys and flow files
+ */
+public class ItemKeys {
```
Since its utility class and it seems to be specific to this new functionality should its visibility be reduced to package private until there is a justified case to make it public?

mans4singh replied:
Changed visibility to package.

mans2singh replied:
Changed visibility to package.

mans2singh commented:
@olegz @jvwing - Please let me know if you have any other recommendations for me.  Thanks

jvwing commented:
@mans2singh - Sorry I'm slow getting back to you.  I have a few comments about the most recent changes:

##### Tests
* The unit tests are much improved.  Your mock implementation is better than my suggested approach, and there has been a big improvement in code coverage as a result.
* I think we could use one or two mocked tests for GetDynamoDB.
* Test coverage for unprocessed items is low, given that your design targets the batch get/put/delete APIs and unprocessed items will be expected.  Earlier today, I thought that would be easily fixed.  It turns out I don't have solid knowledge of what unprocessed responses will really look like in practice, and the AWS SDK is more confusing than helpful.

##### AbstractWriteDynamoDBProcessor 
AbstractWriteDynamoDBProcessor has two methods - handleUnprocessedPutItems and handleUnprocessedDeleteItems - and two subclasses, PutDynamoDB and DeleteDynamoDB.  I know that isn't how things started a couple weeks ago, it used to be shared code, but isn't this a sign that we don't need an AbstractWriteDynamoDBProcessor?  Can we just move those methods to their respective processors, or do you anticipate sharing them with additional writing processors?

##### Relationships
What types of relationships have you considered besides "success" and "failure"?   A malformed input "failure" requires correction before retrying, but unprocessed item failures might simply be retried.  Also, GetDynamoDB might have a separate relationship for "not found" as opposed to "failure".  On the good side, your inclusion of complete response codes in the output attributes makes it possible to filter out the various failure modes.  Additional relationships could be added as features later.

I am still working on setting up some tests for unprocessed items.

mans2singh commented:
@jvwing - Thanks or your feedback.  

Based on your recommendations - I've added separate relations for  unprocessed and not found (latter only for get dynamo processor).  I've also added unit test cases for not found and updated integration/unit tests for unprocessed scenarios for all the processors.  While adding unit test cases for not found a bug which was hiding the error in a failure and I've corrected that.  
I've also refactored the handle unprocessed items to keep only the common code between delete and put processors in the AbstractWriteDynamoDBProcessor.
Please let me know if you have any other recommendations.
Thanks

mans2singh commented:
I've done a clean build locally and it passes and also tested the case which is failing in travis-ci and it passes.  Also, travis is showing different error for jdk 8 vs jdk 7.  

For jdk 8 - travis ci shows the following error (https://s3.amazonaws.com/archive.travis-ci.org/jobs/114105384/log.txt)
Results :

Failed tests: 
  TestListenRELP.testBatching:126->run:207 expected:<3> but was:<1>

For jdk 7 it shows the following error (https://s3.amazonaws.com/archive.travis-ci.org/jobs/114105387/log.txt)

Failed tests: 
  TestListenRELP.testBatching:126->run:207 expected:<3> but was:<2>

Please let me know if there is anything I can do to resolve this failure.

bbende commented:
@mans2singh I think a commit I made on Friday exposed a race condition in TestListenRELP that was always there, but we were getting lucky before and it was never happening. I believe I have this fixed in the PR for adding the Splunk bundle which should be merged in soon and should then make this pass. Sorry for any confusion it created.

jvwing commented:
I ran into a NullPointerException testing GetDynamoDB with a single unprocessed item consisting of a hash key, but no range key.  The NullPointerException is thrown from AbstractDynamoDBProcessor.getAttributeValue() on line 207 when it tries to use the range key AttributeValue, which is null.  I copy/paste/modified one of your tests to illustrate, [308d60cf6167b209d0d222bb6a3e1cff25b1e4cd](https://github.com/jvwing/nifi/commit/308d60cf6167b209d0d222bb6a3e1cff25b1e4cd).

mans2singh commented:
@jvwing - I missed the case where the attribute would be null - thanks for catching that. I've corrected the issue and included your  test case among unit tests.
Please let me know if you have any other recommendation.

mans2singh commented:
@jvwing - I've added more unit test cases for different types of exceptions.  Let me know what you think.

jvwing commented:
I have [another annoying test for you](https://github.com/jvwing/nifi/commit/95d7620f45398f20f23d6f5e7cfe572616beb8d6), for the following scenario:

1. GetDynamoDB successfully receives an item
1. The item does not have the Json Document attribute expected from the configured properties
1. GetDynamoDB throws a NullPointerException with no helpful explanation.  Logged text was:

> 2016-03-08 10:48:40,855 ERROR [Timer-Driven Process Thread-5] o.a.n.p.aws.dynamodb.GetDynamoDB GetDynamoDB[id=86f82c25-0021-4228-a5a6-df7d11c2370e] Could not process flowFiles due to exception : null

I expect this scenario to be fairly normal with the Json Document attribute design since DynamoDB has no schema enforcement.  I recommend changing the code to either throw a more helpful exception message (see sample), or to silently accept this by passing empty FlowFile content.

jvwing commented:
I also have a laundry list of nitpicking comments:

* The names of the relationships for GetDynamoDB are `failure`, `success`, `unprocessed`, and `Not Found`.  I recommend we change proper case "Not Found" to lowercase "not found" to fit in.

* Description of REL_UNPROCESSED in AbstractDynamoDBProcessor does not clarify why items are unprocessed.  I recommend something like:

 > FlowFiles are routed to this relationship when DynamoDB does not process them in the batch.  Typical reasons are insufficient table throughput capacity and exceeding the maximum bytes per request.  Unprocessed FlowFiles are expected to be retry-able without modification.

* `@Ignore` was commented out on ITPutGetDeleteGetDynamoDBTest.java, line 31, so the integration tests are trying to run.  I think this is bothering TravisCI, when it's not running out of memory.

* GetDyanmoDB @CapabilityDescription does not specify where the data goes.  Also, there is a misspelling "parimary".  I recommend something like the following:
 > Retrieves a document from DynamoDB based on hash and range key. The key can be string or number. For any get request all the primary keys are required (hash or hash and range based on the table keys).  A Json Document ("Map") attribute of the DynamoDB item is read into the content of the FlowFile.

* PutDynamoDB @CapabilityDescription, same issue.  I recommend:
 > Puts a document to DynamoDB based on hash and range key. The table can have either hash and range or hash key alone. Currently the keys supported are string and number and value can be json document. In case of hash and range keys both key are required for the operation. The FlowFile content must be JSON.  FlowFile content is mapped to the specified Json Document attribute in the DynamoDB item.

* Commit 51448fb introduced a single-space whitespace inconsistency in GetDynamoDB.java, line 114 "final String jsonDocument = ..."

Yeah, I think we're down to the whitespace issues.  This is looking pretty good to me.

mans2singh commented:
@jvwing - I've add check for empty json response and it will continue to process without an error. I've also incorporated your comments about description in the code.  Let me know what you think.  Thanks

jvwing commented:
Thanks for the updates, @mans2singh , I think this pull request looks good (and works good).

apiri commented:
@mans2singh Would you be able to rebase this to current master and/or get it squashed down into a single commit?  Ran into a number of conflicts trying to get applied to current master and don't want to make any assumptions about the merges.

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/dynamodb/AbstractDynamoDBProcessor.java:
```diff
@@ -0,0 +1,337 @@
+            .name("Hash Key Value Type")
+            .required(true)
+            .description("The hash key value type of the item")
+            .defaultValue("string")
```
For the sake of consistency, I think you should prefer to use ALLOWABLE_VALUE_STRING.getValue() for the default

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/dynamodb/AbstractDynamoDBProcessor.java:
```diff
@@ -0,0 +1,337 @@
+            .name("Range Key Value Type")
+            .required(true)
+            .description("The range key value type of the item")
+            .defaultValue("string")
```
ALLOWABLE_VALUE_STRING.getValue() for the default

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/dynamodb/AbstractDynamoDBProcessor.java:
```diff
@@ -0,0 +1,337 @@
+            .defaultValue(Charset.defaultCharset().name())
+            .build();
+
+    protected DynamoDB dynamoDB;
```
should be marked volatile

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/dynamodb/AbstractDynamoDBProcessor.java:
```diff
@@ -0,0 +1,337 @@
+     */
+    @Override
+    protected AmazonDynamoDBClient createClient(final ProcessContext context, final AWSCredentialsProvider credentialsProvider, final ClientConfiguration config) {
+        getLogger().info("Creating client with credentials provider");
```
might want to drop this down to debug level

mans2singh commented:
I've addressed the minor comments (log level + default value) and will work on rebasing.

mans2singh commented:
@apiri - I've rebased the branch to the current master and have resolve the conflicts.  However, on rebasing there were many updates from the master so I could not squash my changes into one commit.  If you see any conflicts, please let me know.  Thanks.

apiri commented:
@mans2singh thanks for getting this updated, will start looking over it tonight/tomorrow

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/dynamodb/DeleteDynamoDB.java:
```diff
@@ -0,0 +1,161 @@
+
+        final String table = context.getProperty(TABLE).getValue();
+
+        final String hashKeyName = context.getProperty(HASH_KEY_NAME).getValue();
```
This property is listed as supporting EL but is not evaluated via #evaluateAttributeExpressions

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/dynamodb/AbstractDynamoDBProcessor.java:
```diff
@@ -0,0 +1,337 @@
+            .name("Range Key Value")
+            .required(false)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .expressionLanguageSupported(true)
```
Is there a reason why only this property supports EL?  Seems like there are others that are similar in nature that would also benefit.  Not overly familiar with DynamoDB, so please let me know any details.

mans2singh replied:
I was trying to make sure that key names and value types for particular table do not change for a particular flow so they don't support EL.  The value can and does change so they support EL. Let me know your thoughts.

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/dynamodb/PutDynamoDB.java:
```diff
@@ -0,0 +1,182 @@
+
+   public static final List<PropertyDescriptor> properties = Collections.unmodifiableList(
+            Arrays.asList(TABLE, HASH_KEY_NAME, RANGE_KEY_NAME, HASH_KEY_VALUE, RANGE_KEY_VALUE,
+                HASH_KEY_VALUE_TYPE, RANGE_KEY_VALUE_TYPE, JSON_DOCUMENT, DOCUMENT_CHARSET, BATCH_SIZE,
```
JSON_DOCUMENT is listed as being retrieved from the AbstractDynamoDBProcessor class.  This doesn't seem to map to Put.  May want to either make the description generic or, likely better, make it a separate property for this processor.

mans2singh replied:
JSON_DOCUMENT is used in the PutDynamoDB processor to insert the document into that particular field in DynamoDB.

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/dynamodb/PutDynamoDB.java:
```diff
@@ -0,0 +1,182 @@
+                continue;
+            }
+
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
```
Byte arrays are here, but seems like there are also limits to document sizes as per http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Limits.html.  Should perform similar checks.

mans2singh replied:
I've added size check and unit tests for this scenario. Let me know if you have any other recommendation.

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/dynamodb/PutDynamoDB.java:
```diff
@@ -0,0 +1,182 @@
+        final String hashKeyValueType = context.getProperty(HASH_KEY_VALUE_TYPE).getValue();
+        final String rangeKeyName = context.getProperty(RANGE_KEY_NAME).getValue();
+        final String rangeKeyValueType = context.getProperty(RANGE_KEY_VALUE_TYPE).getValue();
+        final String jsonDocument = context.getProperty(JSON_DOCUMENT).getValue();
```
Attribute is okay, but think that having the ability to read from content would also be very useful for writing to Dynamo.

mans2singh replied:
Can you please elaborate on this ?  Thanks

apiri replied:
@mans2singh It looks like this processor only takes json content from the attribute of a FlowFile.  Do you think there is also value in reading the document from the FlowFile's content as well?

mans2singh replied:
@aldrin - The content is taken from the flow file body and the json_document attribute defines the table attribute name where the json content is to be inserted.  Please let me know if you have any other recommendation.

Paul424 replied:
@mans2singh  @apiri I'm evaluating the PutDynamoDB processor but cannot get it to work. I don't see how i can specify the value of the primary (partition) key using the HASH_KEY_VALUE setting. How do i point to a given field in the flowfile record? The default value ${dynamodb.item.hash.key.value} does not make sense to me?
Any help is appreciated,
Paul

apiri replied:
Hey @Paul424!

Sorry for missing your previous question, my filters consumed the notification email, but saw that you got a resolution and your approach sounds great.

The key thing to underscore is that a FlowFile is a set of attributes and a content section.  Expression Language (EL) is a way to do some quick operations on those attributes.  Attributes are effectively metadata about the content and aid in making decisions and characterizing the data upfront so that the content can make its way through the appropriate pipeline and processing. 

If you happen to run into anything else in terms of AWS processors, or even NiFi in general, feel free to reach out to the community.  (http://nifi.apache.org/mailing_lists.html)  Any feedback on how we could make this or anything else easier is always greatly appreciated!

mans2singh commented:
@apiri @jvwing  - Please let me know if you have any further recommendations.  Thanks

mans2singh commented:
@apiri - Hi Aldrin - Do you have any recommendations for this pull request ?  Thanks

apiri commented:
@mans2singh sorry for the delays.  Going to revisit this now so we can square things away.

apiri commented:
Looks good.  Was able to verify functionality.  Merged into master and 0.x.

Thanks @mans2singh!

mans2singh commented:
Thanks @aldrin @jvwing @olegz @bbende for your advice/guidance in getting this feature implemented.

apiri commented:
@mans2singh No problem!  Sorry for the delays in getting incorporated but appreciate you bearing with us and all the great contributions to the AWS side of the house.  👍

davetorok commented:
Hi @mans2singh we desperately need support for the ENDPOINT_OVERRIDE property in all the DynamoDB processors, similar to the S3 processors.  It is probably as simple as 
1 - adding ENDPOINT_OVERRIDE to the properties Collections arrays in all the processors
2 - adding in the createClient  methods in AbstractDynamoDBProcessor.java:
`String endpointurl = (StringUtils.trimToEmpty(context.getProperty(ENDPOINT_OVERRIDE).getValue());
if (endpointurl.isEmpty() == false){ client.setEndpoint(endpointurl}`

apiri commented:
Hey @davetorok!

Seems like a reasonable request.   I created a JIRA (https://issues.apache.org/jira/browse/NIFI-1855) to track this.  Please feel free to add any additional thoughts you might have there.
